quinn-proto/src/connection.rs

```diff
@@ -1923,6 +1923,253 @@ impl Connection {
+            let crypto = if let Some(ref crypto) = space.crypto {
+                crypto
+            } else if space_id == SpaceId::Data {
+                self.zero_rtt_crypto.as_ref().unwrap()
+            } else {
+                unreachable!("tried to send {:?} packet without keys", space_id);
+            };
+
+            let mut padded = if self.side.is_client() && space_id == SpaceId::Initial {
+                // Initial-only packets MUST be padded
+                buf.resize(MIN_INITIAL_SIZE - crypto.packet.tag_len(), 0);
+                true
+            } else {
+                false
+            };
+
+            let pn_len = number.len();
+            // To ensure that sufficient data is available for sampling, packets are padded so that the
+            // combined lengths of the encoded packet number and protected payload is at least 4 bytes
+            // longer than the sample required for header protection.
+            let protected_payload_len = (buf.len() + crypto.packet.tag_len())
+                - partial_encode.start
+                - partial_encode.header_len;
+            if let Some(padding_minus_one) =
+                (crypto.header.sample_size() + 3).checked_sub(pn_len + protected_payload_len)
+            {
+                let padding = padding_minus_one + 1;
+                padded = true;
+                trace!(self.log, "PADDING * {count}", count = padding);
+                buf.resize(buf.len() + padding, 0);
+            }
+
+            buf.resize(buf.len() + crypto.packet.tag_len(), 0);
+            let packet_buf = &mut buf[partial_encode.start..];
+            partial_encode.finish(
+                packet_buf,
+                &crypto.header,
+                Some((exact_number, &crypto.packet)),
+            );
+
+            if let Some((sent, acks)) = sent {
+                // If we sent any acks, don't immediately resend them. Setting this even if ack_only is
+                // false needlessly prevents us from ACKing the next packet if it's ACK-only, but saves
+                // the need for subtler logic to avoid double-transmitting acks all the time.
+                space.permit_ack_only &= acks.is_empty();
+
+                self.on_packet_sent(
+                    now,
+                    space_id,
+                    exact_number,
+                    SentPacket {
+                        acks,
+                        time_sent: now,
+                        size: if padded || !ack_only {
+                            buf.len() as u16
+                        } else {
+                            0
+                        },
+                        is_crypto_packet: space_id != SpaceId::Data && !sent.crypto.is_empty(),
+                        ack_eliciting: !ack_only,
+                        retransmits: sent,
+                    },
+                );
+            }
+
+            if !coalesce || buf.capacity() - buf.len() < 40 {
```

Comment:
Can we make `40` not a magic number here?

Reply:
Sure, what do you want to call it? `MIN_COALESCING_SIZE`? I just picked a value at random.

Reply:
`MIN_PACKET_SPACE` maybe?

Reply:
Much better.